expression/visitor.go

```diff
@@ -0,0 +1,396 @@
+	VisitPatternIn(n *PatternIn) (Expression, error)
+
+	// VisitPatternLike visits PatternLike expression.
+	VisitPatternLike(p *PatternLike) (Expression, error)
+
+	// VisitPatternRegexp visits PatternRegexp expression.
+	VisitPatternRegexp(p *PatternRegexp) (Expression, error)
+
+	// VisitPExpr visits PExpr expression.
+	VisitPExpr(p *PExpr) (Expression, error)
+
+	// VisitPosition visits Position expression.
+	VisitPosition(p *Position) (Expression, error)
+
+	// VisitRow visits Row expression.
+	VisitRow(r *Row) (Expression, error)
+
+	// VisitSubQuery visits SubQuery expression.
+	VisitSubQuery(sq SubQuery) (Expression, error)
+
+	// VisitUnaryOperation visits UnaryOperation expression.
+	VisitUnaryOperation(u *UnaryOperation) (Expression, error)
+
+	// VisitValue visits Value expression.
+	VisitValue(v Value) (Expression, error)
+
+	// VisitValues visits Values expression.
+	VisitValues(v *Values) (Expression, error)
+
+	// VisitVariable visits Variable expression.
+	VisitVariable(v *Variable) (Expression, error)
+
+	// VisitWhenClause visits WhenClause expression.
+	VisitWhenClause(w *WhenClause) (Expression, error)
+}
+
+// BaseVisitor is the base implementation of Visitor.
+// It traverse the expression tree and call expression's Accept function.
+// It can not be used directly.
+// A specific Visitor implementation can embed it in and only implement
+// desired methods to do the job.
+type BaseVisitor struct {
+	V Visitor
+}
+
+// VisitBetween implements Visitor interface.
+func (bv *BaseVisitor) VisitBetween(b *Between) (Expression, error) {
+	var err error
+	b.Expr, err = b.Expr.Accept(bv.V)
+	if err != nil {
+		return b, errors.Trace(err)
+	}
+	b.Left, err = b.Left.Accept(bv.V)
+	if err != nil {
+		return b, errors.Trace(err)
+	}
+	b.Right, err = b.Right.Accept(bv.V)
+	if err != nil {
+		return b, errors.Trace(err)
+	}
+	return b, nil
+}
+
+// VisitBinaryOperation implements Visitor interface.
+func (bv *BaseVisitor) VisitBinaryOperation(o *BinaryOperation) (Expression, error) {
+	var err error
+	o.L, err = o.L.Accept(bv.V)
+	if err != nil {
+		return o, errors.Trace(err)
+	}
+	o.R, err = o.R.Accept(bv.V)
+	if err != nil {
+		return o, errors.Trace(err)
+	}
+	return o, nil
+}
+
+// VisitCall implements Visitor interface.
+func (bv *BaseVisitor) VisitCall(c *Call) (Expression, error) {
+	var err error
+	for i := range c.Args {
+		c.Args[i], err = c.Args[i].Accept(bv.V)
+		if err != nil {
+			return c, errors.Trace(err)
+		}
+	}
+	return c, nil
+}
+
+// VisitCompareSubQuery implements Visitor interface.
+func (bv *BaseVisitor) VisitCompareSubQuery(cs *CompareSubQuery) (Expression, error) {
+	var err error
+	cs.L, err = cs.L.Accept(bv.V)
+	if err != nil {
+		return cs, errors.Trace(err)
+	}
+	_, err = cs.R.Accept(bv.V)
+	if err != nil {
+		return cs, errors.Trace(err)
+	}
+	return cs, nil
+}
+
+// VisitDefault implements Visitor interface.
+func (bv *BaseVisitor) VisitDefault(d *Default) (Expression, error) {
+	return d, nil
```

Comment:
Default doesn't have children expressions, so we don't need to do anything here.